File: android/app/src/main/java/com/example/smartplanner/data/NewsService.kt
package com.example.smartplanner.data

import com.example.smartplanner.model.NewsArticle
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.time.LocalDate
import java.time.format.DateTimeFormatter

object NewsService {

    fun fetchTopStories(): List<NewsArticle> {
        val connection = (URL(NewsApiConfig.MOST_POPULAR_URL).openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 15000
            readTimeout = 15000
        }

        try {
            val responseCode = connection.responseCode
            if (responseCode !in 200..299) {
                error("Unexpected response code: $responseCode")
            }

            val payload = connection.inputStream.bufferedReader().use { it.readText() }
            val results = JSONObject(payload).getJSONArray("results")

            return buildList {
                for (index in 0 until results.length()) {
                    val article = results.getJSONObject(index)
                    add(
                        NewsArticle(
                            title = article.optString("title"),
                            abstractText = article.optString("abstract"),
                            publishedDate = formatDate(article.optString("published_date")),
                            source = article.optString("source"),
                            section = article.optString("section"),
                            imageUrl = extractImageUrl(article)
                        )
                    )
                }
            }
        } finally {
            connection.disconnect()
        }
    }

    private fun extractImageUrl(article: JSONObject): String? {
        val mediaList = article.optJSONArray("media") ?: return null
        for (index in 0 until mediaList.length()) {
            val media = mediaList.optJSONObject(index) ?: continue
            val metadata = media.optJSONArray("media-metadata") ?: continue
            for (metaIndex in metadata.length() - 1 downTo 0) {
                val entry = metadata.optJSONObject(metaIndex) ?: continue
                val url = entry.optString("url")
                if (url.isNotBlank()) {
                    return url
                }
            }
        }
        return null
    }

    private fun formatDate(value: String): String {
        return runCatching {
            LocalDate.parse(value).format(DateTimeFormatter.ofPattern("dd.MM.yyyy"))
        }.getOrDefault(value)
    }
}

File: ios/SmartPlanner/Services/NewsService.swift
import Foundation

protocol NewsServiceProtocol {
    func fetchTopStories() async throws -> [NewsArticle]
}

struct NewsService: NewsServiceProtocol {
    private let session: URLSession
    private let inputDateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        formatter.locale = Locale(identifier: "en_US_POSIX")
        return formatter
    }()
    private let outputDateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "dd.MM.yyyy"
        return formatter
    }()

    init(session: URLSession = .shared) {
        self.session = session
    }

    func fetchTopStories() async throws -> [NewsArticle] {
        let (data, response) = try await session.data(from: NewsAPIConfig.mostPopularURL)
        guard let httpResponse = response as? HTTPURLResponse, 200..<300 ~= httpResponse.statusCode else {
            throw URLError(.badServerResponse)
        }

        let decoder = JSONDecoder()
        let payload = try decoder.decode(MostPopularResponse.self, from: data)

        return payload.results.map { article in
            NewsArticle(
                id: "\(article.title)-\(article.publishedDate)",
                title: article.title,
                abstract: article.abstract,
                publishedAt: formatDate(article.publishedDate),
                source: article.source,
                section: article.section,
                imageURL: article.media?.first?.metadata.last?.url
            )
        }
    }

    private func formatDate(_ value: String) -> String {
        guard let date = inputDateFormatter.date(from: value) else { return value }
        return outputDateFormatter.string(from: date)
    }
}
